Community Translator V2: Load user waiting translations

deBhal opened the pull request:
#### Changes proposed in this Pull Request

This PR loads and displays current and waiting translations from the user, so if they go to the trouble of submitting a translation, they don't have to keep looking at the old ones.

Still TODO:

- [x] Move the visibility logic out of `layout`

#### Testing instructions

All the changes in this PR are behind the `i18n/community-translator` feature flag, so you'll want to start up calypso with `ENABLE_FEATURES=i18n/community-translator npm start`

For testing, we should set the GP project to 'test' (in `loadUndeployedTranslations()` in `client/lib/i18n-utils/switch-locale.js`) and add translations there, but it's ok to submit a few strings for testing - just remember to reject them later.

Once you've got some strings under your username, turn on the community translator, and check that they're picked up:

![Mon_profil_—_WordPress_com_and_Translations___French__Canada____WordPress_com___GlotPress_-_julesaus_dev_dfw_wordpress_com](https://user-images.githubusercontent.com/5952255/61206924-1e9a9d80-a737-11e9-9a32-5cd6bb5bdac1.jpg)


<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Commit:
load user waiting translations

Commit:
Reduxify & simplify translator

Commit:
update layout when translator enabled status becomes available

jsnajdr commented on client/components/community-translator/index.jsx:
```diff
@@ -158,8 +171,13 @@ class CommunityTranslator extends Component {
+	translatorEnabled: getUserSetting( state, ENABLE_TRANSLATOR_KEY ),
+} );
+
+export default connect( mapState )( localize( CommunityTranslator ) );
```
Two nits: the component doesn't seem to use the `localize` HOC, and using `PureComponent` is not necessary here: all Redux-connected components are pure by default.

deBhal replied:
> the component doesn't seem to use the localize HOC,

I made a comment about this in the component, which I've moved down to the localize() call itself.

Even though we don't use `translate()` we still need this component to update when the translations change. In particular, we need to redraw to pick up changes when i18n finishes loading new data.

This has made the rerendering logic pretty complicated overall, and I've got to where it's at now by incrementally fixing failing updates. I think it's right now in the sense that it works, but I wonder if there's a conceptually better approach that I'm not seeing.

> all Redux-connected components are pure by default.

Thanks, :TIL: https://react-redux.js.org/api/connect#pure-boolean

jsnajdr commented on client/layout/index.jsx:
```diff
@@ -223,5 +228,8 @@ export default connect( state => {
 		shouldQueryAllSites: currentRoute && currentRoute !== '/jetpack/connect/authorize',
+		communityTranslatorEnabled:
+			config.isEnabled( 'i18n/community-translator' ) &&
+			getUserSetting( state, ENABLE_TRANSLATOR_KEY ),
```
Can we help make the `Layout` component more clean and generic by moving all the community translator logic out of it? The async loading could be done like this:

In `Layout`, we ideally do just:
```jsx
<CommunityTranslator />
```

Then `CommunityTranslator` checks all the conditions and optionally async loads larger subcomponents:
```jsx
function CommunityTranslator( { isEnabled } ) {
  if ( isEnabled && someConditionHolds ) {
    return <AsyncLoad require="translator/v1/launcher" />;
  } else if ( anotherConditionHolds ) {
    return <AsyncLoad require="translator/v2/new-shiny-impl" />;
  } else {
    return null;
  }
}

connect( getUserSettingsEtcFromRedux )( CommunityTranslator );
```

jsnajdr reviewed (commented):
All the changes in this PR are active only when the `i18n/community-translator` flag is enabled, which it isn't in any environment. Am I supposed to enable it manually for testing? Bit confused about this.

Commit:
rework flows to ensure successful startup

deBhal commented:
Rebased

Commit:
fix rebase

Commit:
Fix rerender reference after rebase

Commit:
We don't need PureComponent with connect() props @jsnajdr

deBhal commented:
> Am I supposed to enable it manually for testing? Bit confused about this.

Yeah, sorry. I've updated the testing instructions.

Commit:
Improve explanatory comments around i18n rerendering

Commit:
rerender all translations when translator is disabled

Commit:
remove unused variables

Commit:
Fix CORS for postRequest()

Commit:
touch up rebase

Commit:
Move visibilty logic out of client/layout

Commit:
drop WIP comment

Commit:
Enable loading waiting translations

deBhal commented:
@jsnajdr: Lots of things happened, but I've rebased and addressed your feedback - I've moved the logic out of layout and switched away from PureComponent.

Can you please have another look?

lancewillett commented:
Cleaning up older pull requests as part of a GitHub scrub.

Is this change still valid? What's needed to move this along?

If not valid, please close with a comment. Thanks!

deBhal commented:
I'm closing this now, as I'm not likely to get back to it.

It's a shame, but someone else could re-open it if it's useful.
